core/src/main/java/com/github/mizool/core/concurrent/FutureStreamJoiner.java

```diff
@@ -0,0 +1,195 @@
+            Stream<ListenableFuture<Void>> listenableFutures, int concurrencyLimit, ExecutorService executorService)
+        {
+            verifyConcurrencyLimit(concurrencyLimit);
+
+            Stream<Void> results = BufferedStreamAdapter.listenable()
+                .adapt(listenableFutures, concurrencyLimit, executorService);
+
+            return (ListenableFuture<Void>) MoreExecutors.listeningDecorator(executorService)
+                .submit(consumeStream(results));
+        }
+    }
+
+    /**
+     * Provides fluent syntax for joining completable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     */
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
+    public class Completable
+    {
+        /**
+         * Returns a future that will complete normally once all futures in the given stream have completed
+         * normally. <br>
+         * <br>
+         * See {@link FutureStreamJoiner} for details.
+         *
+         * @param completableFutures the futures to join
+         * @param concurrencyLimit the maximum number of incomplete futures at any point in time
+         * @param executorService where to execute the stream consumer
+         */
+        public CompletableFuture<Void> join(
+            Stream<CompletableFuture<Void>> completableFutures, int concurrencyLimit, ExecutorService executorService)
+        {
+            verifyConcurrencyLimit(concurrencyLimit);
+
+            Stream<Void> results = BufferedStreamAdapter.completable()
+                .adapt(completableFutures, concurrencyLimit, executorService);
+
+            return CompletableFuture.runAsync(consumeStream(results), executorService);
+        }
+    }
+
+    /**
+     * Entry point for joining listenable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     *
+     * @return fluent syntax
+     */
+    public Listenable listenable()
+    {
+        return new Listenable();
+    }
+
+    /**
+     * Entry point for joining completable futures. <br>
+     * <br>
+     * See {@link FutureStreamJoiner} for details.
+     *
+     * @return fluent syntax
+     */
+    public Completable completable()
+    {
+        return new Completable();
+    }
+
+    private void verifyConcurrencyLimit(int concurrencyLimit)
+    {
+        if (concurrencyLimit <= 0)
+        {
+            throw new IllegalArgumentException("Concurrency limit must be greater than 0");
+        }
+    }
+
+    private Runnable consumeStream(Stream<Void> stream)
+    {
+        return new StreamConsumingWorker(stream);
+    }
+
+    @Value
+    @Getter(value = AccessLevel.NONE)
+    private class StreamConsumingWorker implements Runnable
+    {
+        Stream<Void> stream;
+
+        @Override
+        @SneakyThrows
+        public void run()
+        {
+            try
+            {
+                stream.forEach(whatever -> {
+                });
+            }
+            catch (UncheckedExecutionException t)
```

Comment:
Reading this, I wonder if the `BufferedStreamAdapter` should even expose the unchecked execution exception.

#### Full stack example
During the store operation, a `Stream<CompletableFuture<FooRecord>>` comes into existence.
This is then parallelized using the `BufferedStreamAdapter`, resulting in a `Stream<FooRecord>`.
This is then mapped to a `Stream<Foo>` and propagated towards the REST layer, where it's again mapped to a `Stream<FooDto>` and then serialized to JSON.

This means, exceptions, that occur during consumption of the stream, happen during serialization. That part of the code is totally unaware about buffering, parallelization and everything that comes with it.

#### Theory
Since the `BufferedStreamAdapter` returns a `Stream<V>`, it also hides that there are futures and threads involved. Thus, it should also hide all exceptions that come from using futures and threads.

#### Idea
Let's centralize handling of execution exceptions inside the `handleFutureResult` method.
A "future result" can either be a value or an exception, as indicated by the signature of the `handleFutureResult` method.
If it's an exception, is definitely _has_ qualities of "something you only get when dealing with futures", either because the original exception from inside the future was wrapped or because there was an exception during execution of the future.
In both cases, after `handleFutureResult` is finished, there should be no trace of "futures" left. Even the `ValueHolder` does not need to know about futures or special exceptions that occur when dealing with futures, as the `ValueHolder` is already positioned _after_ the future was finished and only helps us to transport a value or a (non-future-related) exception towards the `BlockingSpliterator`.

Based on that, this `forEach` (and all other terminal operations on streams, for that matter) would not have to deal with an `UncheckedExecutionException`. Why would a stream even throw such a thing? Abstruse.

Reply:
You certainly make a good point here.

To sum this up and put it another way: the stream returned by `BufferedStreamAdapter` should, for all intents and purposes, feel like a stream that produces values in a **synchronous** way, just as if some algorithm would compute them. And exceptions in such computations would only ever throw regular, non-futurish exceptions - the worst that could happen to a stream consumer is that it encounters a `RuntimeException` (or subclass) wrapping a checked exception.

That said, this wrapping is the officially documented behavior of `BufferedStreamAdapter`. Thus, I'm hesitant to change it in a backwards-incompatible way (we may be allowed to do so by semantic versioning because we're not 1.0 yet, but still). Also note that `FutureStreamJoiner`, despite being built on top of `BufferedStreamAdapter`, would not be affected as the joint future it produces is specified (and unit-tested) to behave exactly like any other future (wrapping exceptions in an `ExecutionException`).

I suggest we carefully analyze all call sites of this class in relevant corporate code bases. I expect that at least some of them do explicit exception checking and unwrapping and could break if the exceptions they want to handle are suddenly thrown directly instead of being wrapped.

Let's discuss this in person, and maybe do the aforementioned analysis.

Also, as a first step, I'll try extending the existing unit test for exception wrapping so we have a solid foundation. Right now, it only throws a runtime exception, not the full suite of throwable kinds that the `FutureStreamJoiner` tests verify. I also suspect inconsistent behavior depending on which element of the stream caused the exception (like the original problem with `ListenableFutureCollector` that led to us writing `FutureStreamJoiner`).

Reply:
You are right: I did not take into account that the wrapped exceptions are not only the current behavior but also the documented behavior. I agree that we should check usages first before breaking the API.
That being said: adapting the call sites to my proposed behavior should generally simplify them, so the changes are most likely pretty easy. But they have to be identified and performed regardless.

Reply:
Since your calendar is the more... complicated one... please schedule a one hour meeting between the two of us for the purpose of checking implications of the proposed change. When that's scheduled, this conversation can be resolved.